source/docs/essentials/extend-the-graphql-schema.md

```diff
@@ -0,0 +1,443 @@
+seconds):
+
+```js
+// my-module/server/modules/clicks-counters/resolvers.js
+const counters = new Map();
+
+const currentValueOf = sku => counters.get(sku) || 0;
+
+module.exports = {
+  Product: {
+    clicksCounter: ({ sku }) => currentValueOf(sku)
+  },
+
+  Mutation: {
+    incrementProductCounter(_, { sku, incrementValue = 1 }) {
+      counters.set(sku, currentValueOf(sku) + incrementValue);
+      return {
+        success: true
+      };
+    }
+  }
+};
+```
+
+Let’s analyze this code part by part to understand what is possible to achieve
+in a module.
+
+First of all, the exported resolver map defines a resolver for the
+`clicksCounter` field of any `Product` field.
+
+```js
+// …
+module.exports = {
+  Product: {
+    clicksCounter: ({ sku }) => currentValueOf(sku)
+  }
+  // …
+};
+```
+
+The resolver function will use the `sku` key from its parent data and will
+return the the current counter value from its local state (defaulting to `0` if
+no clicks occured).
+
+It is important to understand that the `sku` is not a parameter provided by
+frontend developers in the Query. It comes from data returned by earlier
+resolvers that fetched a `Product`, no matter where in the graph (category,
+upsells, cart…).
+
+<blockquote class="info">
+Please note that it could have been implemented as a top level Query as well:
+```graphql
+extend type Query {
+  clicksCounterByProductSKU(sku: String!): Int
+}
+```
+However we feel it would have been less intuitive for frontend developers. It
+may seem more natural for people used to design relational databases or REST
+APIs, but we recommend to try limiting as much as possible the number of
+top-level queries in your projects to learn _thinking in GraphQL_.
+
+</blockquote>
+
+The second part of this resolver map is a `incrementProductCounter` mutation.
+Mutations in GraphQL are a way to modify server-side data (whereas queries allow
+to read data). GraphQL modules must declare mutations by extending the top-level
+GraphQL `Mutation` type. You can read more about Mutations in the
+[GraphQL’s Mutations documentation](https://graphql.org/learn/queries/#mutations).
+
+```js
+// …
+module.exports = {
+  // …
+  Mutation: {
+    incrementProductCounter(_, { sku, incrementValue = 1 }) {
+      counters.set(sku, currentValueOf(sku) + incrementValue);
+      return {
+        success: true
+      };
+    }
+  }
+};
+```
+
+In this resolver, the first parameter is unused (because in a top level resolver
+no earlier resolvers have been run so there are no _parent data_). The second
+parameter contains arguments passed to the mutation. These parameters are the
+`(sku: String!, incrementValue: Int)` part of the schema definition declared
+earlier:
+
+```graphql
+extend type Mutation {
+  incrementProductCounter(sku: String!, incrementValue: Int): MutationSuccess
+}
+```
+
+The resolver uses the mandatory `sku` argument to fetch the current counter
```

Comment:
This paragraph feels a bit confusing to me. I don't really know how to express it in a better way.

Reply:
I’ll try with bullet points.